classes/book.lua

```diff
@@ -100,6 +93,7 @@ end
 function book:finish ()
   local ret = plain.finish(self)
   self:writeToc()
+  self:writeLabelRefs()
```

Comment:
This brings up an interesting point ... not really the fault of this package contribution but a larger fault in our current core implementation: it really should not be necessary to add code to the base class to get stuff to happen in these locations. We need more hook/callback functions so that packages can get their work done without needing core or class adjustments.

Reply:
On the other hand, there could be cases where the class knows more regarding when and what to do (e.g. in a different order, at some other point, etc.). Actually we could need to distinguish real '"packages" (that provide extra commands without needing class support) and "modules" that are somewhat reusable in various classes but depend on them. I am not fully sure yet, but it could be opened for discussion. (But it's true that the current way is a bit weird and even the naming is inconsistent, e.g. writeTocNodes, writeLabelRefs, buildIndex)

Reply:
Yes, order is an issue. In a hook system probably that would come down to order in which modules were first loaded. Given there are a couple stages involved that could be hooked, it may not even be pertinent to need an order, and it if was whatever package cared about the other one could just make sure it was loaded first.

Perhaps the hooks would need to be tagged by package so that, for example, a cross-ref package could request that (if present) any toc package hooks be called before it's own...

Reply:
Now using the "finish" hook.